### PR TITLE
[feature](delete-predicate) support delete sub predicate v2

### PR DESCRIPTION
**Pull request by Hastyshell**
## Proposed changes

1. New structure for delete sub predicate.

Delete sub predicate uses a string type `condition_str` to stored temporarily now and fields will be extracted from it using std::regex, which may introduces stack overflow when matching a extremely large string(bug of libc).

Now we attempt to use a new PB structure to hold the delete sub predicate, to avoid that problem.

```proto
message DeleteSubPredicatePB {
    optional int32 column_unique_id = 1;
    optional string column_name = 2;
    optional string op = 3;
    optional string cond_value = 4;
}
```

Currently, 2 versions of sub predicate will both be filled. For query, we use the v2, and during compaction we still use v1. The old rowset meta with delete predicates which had sub predicate v1 will be attempted to convert to v2 when read from PB. Moreover, efforts will be made to rewrite these meta with the new delete sub predicate.

2. Make preparation to use column unique id to specify a column globally.

Using the column unique id rather than the column name to identify a column is vital for flexible schema change. The rewritten delete predicate will attach column unique id.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.2 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162305888 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.62 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17162311609 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.66 seconds
 stream load tsv:          504 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162340743 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.71 seconds
 stream load tsv:          504 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162489471 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.93 seconds
 stream load tsv:          517 seconds loaded 74807831229 Bytes, about 137 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162330020 Bytes

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.94 seconds
 stream load tsv:          514 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162495701 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.89 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162490744 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.77 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.6 seconds inserted 10000000 Rows, about 337K ops/s
 storage size: 17162266646 Bytes

File: be/src/olap/delete_handler.cpp [hunk 1]
```diff
@@ -81,35 +82,36 @@ Status DeleteHandler::generate_delete_predicate(const TabletSchema& schema,
             LOG(INFO) << "store one sub-delete condition. condition name=" << in_pred->column_name()
                       << "condition size=" << in_pred->values().size();
         } else {
-            string condition_str = construct_sub_predicates(condition);
-            del_pred->add_sub_predicates(condition_str);
-            LOG(INFO) << "store one sub-delete condition. condition=" << condition_str;
+            DeleteSubPredicatePB* sub_predicate = del_pred->add_sub_predicates_v2();
```
Comment (yiguolei) [inline be/src/olap/delete_handler.cpp:85]: 你这个改法可能有问题。
当一个表没开启light weight schema change的时候，这里感觉还是得按照老的来

Comment (TangSiyang2001) [reply be/src/olap/delete_handler.cpp:85]: 诶但是现在不开lsc没问题，开完以后可能会core

File: be/src/olap/delete_handler.cpp [hunk 2]
```diff
@@ -203,7 +205,20 @@ Status DeleteHandler::check_condition_valid(const TabletSchema& schema, const TC
     return Status::OK();
 }
 
-bool DeleteHandler::_parse_condition(const std::string& condition_str, TCondition* condition) {
+Status DeleteHandler::_parse_condition(const DeleteSubPredicatePB& sub_cond,
```
Comment (yiguolei) [inline be/src/olap/delete_handler.cpp:208]: 1. compaction ---> read delete predicates --> v1
2. read query ---> read delete predicates ---> v2
3. read old delete predicates ---> convert v1 to v2

File: gensrc/proto/olap_file.proto [hunk 3]
```diff
@@ -142,6 +142,13 @@ message DeletePredicatePB {
     required int32 version = 1; // This field is useless, but could not removed, not depend on it
     repeated string sub_predicates = 2;
     repeated InPredicatePB in_predicates = 3;
+    repeated DeleteSubPredicatePB sub_predicates_v2 = 4;
+}
+
+message DeleteSubPredicatePB {
+    optional string column_name = 1;
+    optional string op = 2;
```
Comment (yiguolei) [inline gensrc/proto/olap_file.proto:150]: column unique id

File: be/src/olap/delete_handler.cpp [hunk 4]
```diff
@@ -259,21 +303,14 @@ Status DeleteHandler::init(TabletSchemaSPtr tablet_schema,
...
-            auto predicate =
-                    parse_to_predicate(tablet_schema, condition, _predicate_arena.get(), true);
-            if (predicate != nullptr) {
-                temp.column_predicate_vec.push_back(predicate);
-            }
+        if (delete_condition.sub_predicates_size() == 0) {
+            RETURN_IF_ERROR(_parse_column_pred(delete_pred_related_schema,
+                                               delete_condition.sub_predicates_v2(), &temp));
+        } else {
+            // make it compatible with the former versions
```
Comment (yiguolei) [inline be/src/olap/delete_handler.cpp:310]: a int, b int

delete where a = 1

rename a --> c

drop column a 

add column a string

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.89 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162220273 Bytes

File: be/src/olap/reader.cpp [hunk 5]
```diff
@@ -616,9 +616,13 @@ Status TabletReader::_init_delete_condition(const ReaderParams& read_params) {
                       ((read_params.reader_type == ReaderType::READER_CUMULATIVE_COMPACTION &&
                         config::enable_delete_when_cumu_compaction)) ||
                       read_params.reader_type == ReaderType::READER_CHECKSUM);
-
-    return _delete_handler.init(_tablet_schema, read_params.delete_predicates,
-                                read_params.version.second);
+    if (_filter_delete) {
+        return _delete_handler.init<false>(_tablet_schema, read_params.delete_predicates,
```
Comment (yiguolei) [inline be/src/olap/reader.cpp:620]: 这种地方就不要用模板了，意义不大

File: be/src/olap/data_dir.cpp [hunk 6]
```diff
@@ -373,6 +375,14 @@ Status DataDir::load() {
                                     const std::string& meta_str) -> bool {
         RowsetMetaSharedPtr rowset_meta(new RowsetMeta());
         bool parsed = rowset_meta->init(meta_str);
+        // convert delete predicate to v2 (#22442)
+        if (rowset_meta->has_delete_predicate()) {
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp:379]: need write back and then we could delete these code in 2.2

Comment (TangSiyang2001) [reply be/src/olap/data_dir.cpp:379]: ok, actually I'm thinking how to write back in one traverse with least change, our rowset meta traverse is under read semantics.

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.57 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17162502361 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run click bench

**Comment by TangSiyang2001** [comment]
run clickbench

**Comment by TangSiyang2001** [comment]
run arm

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run compile

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run p1

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.01 seconds
 stream load tsv:          540 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162313091 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.71 seconds
 stream load tsv:          543 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17161961989 Bytes

**Comment by TangSiyang2001** [comment]
run feet

**Comment by TangSiyang2001** [comment]
run feut

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.67 seconds
 stream load tsv:          541 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162226175 Bytes

File: be/src/olap/rowset/rowset_meta_manager.cpp [hunk 7]
```diff
@@ -443,25 +443,51 @@ Status RowsetMetaManager::traverse_rowset_metas(
...
-        TabletUid tablet_uid(uid_parts[0], uid_parts[1]);
-        return func(tablet_uid, rowset_id, value);
+        return func(res.first, res.second, value);
     };
     Status status =
             meta->iterate(META_COLUMN_FAMILY_INDEX, ROWSET_PREFIX, traverse_rowset_meta_func);
     return status;
 }
 
+Status RowsetMetaManager::traverse_rowset_metas_with_write(
```
Comment (yiguolei) [inline be/src/olap/rowset/rowset_meta_manager.cpp:457]: 不要写这个函数，这个函数不会长期存在，你写完这个功能后，这个函数很快就要删掉。

File: be/src/olap/data_dir.cpp [hunk 8]
```diff
@@ -381,10 +382,19 @@ Status DataDir::load() {
         if (rowset_meta->is_local()) {
             rowset_meta->set_fs(local_fs);
         }
+        if (rowset_meta->has_delete_predicate()) {
+            auto* delete_pred = rowset_meta->mutable_delete_pred_pb();
+            if (!delete_pred->sub_predicates().empty() &&
+                delete_pred->sub_predicates_v2().empty()) {
+                DeleteHandler::convert_to_sub_pred_v2(delete_pred, rowset_meta->tablet_schema());
+                rowset_meta->serialize(result);
+            }
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp:391]: 直接在这里写回到olap meta 里

File: be/src/olap/data_dir.cpp [hunk 9]
```diff
@@ -381,10 +382,19 @@ Status DataDir::load() {
         if (rowset_meta->is_local()) {
             rowset_meta->set_fs(local_fs);
         }
+        if (rowset_meta->has_delete_predicate()) {
+            auto* delete_pred = rowset_meta->mutable_delete_pred_pb();
+            if (!delete_pred->sub_predicates().empty() &&
+                delete_pred->sub_predicates_v2().empty()) {
+                DeleteHandler::convert_to_sub_pred_v2(delete_pred, rowset_meta->tablet_schema());
+                rowset_meta->serialize(result);
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp:390]: 这里要加log，万一出错我们可以补救

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.18 seconds
 stream load tsv:          543 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162082784 Bytes

File: be/src/olap/reader.cpp [hunk 10]
```diff
@@ -616,9 +616,19 @@ Status TabletReader::_init_delete_condition(const ReaderParams& read_params) {
...
+        // note(tsy): for compaction, keep delete sub pred v1 temporarily
+        return _delete_handler.init(_tablet_schema, read_params.delete_predicates,
+                                    read_params.version.second, false);
+    }
+    auto* runtime_state = read_params.runtime_state;
+    // note(tsy): for query, use session var to enable delete sub pred v2, for schema change, use v2 directly
+    bool enable_sub_pred_v2 =
+            runtime_state == nullptr
+                    ? true
+                    : runtime_state->query_options().enable_delete_sub_predicate_v2;
```
Comment (yiguolei) [inline be/src/olap/reader.cpp:629]: add a method in runtime state ---  enable_delete_predicate_v2()

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.48 seconds
 stream load tsv:          553 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162327719 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.43 seconds
 stream load tsv:          546 seconds loaded 74807831229 Bytes, about 130 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162062334 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.3 seconds
 stream load tsv:          559 seconds loaded 74807831229 Bytes, about 127 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17162181881 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.23 seconds
 stream load tsv:          559 seconds loaded 74807831229 Bytes, about 127 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17161783037 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.1 seconds
 stream load tsv:          554 seconds loaded 74807831229 Bytes, about 128 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.7 seconds inserted 10000000 Rows, about 336K ops/s
 storage size: 17162240523 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.98 seconds
 stream load tsv:          549 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.6 seconds inserted 10000000 Rows, about 337K ops/s
 storage size: 17162015323 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run clickbench

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.2 seconds
 stream load tsv:          550 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162088510 Bytes

File: be/src/runtime/runtime_state.h [hunk 11]
```diff
@@ -434,6 +434,11 @@ class RuntimeState {
                        : 0;
     }
 
+    inline bool enable_sub_pred_v2() const {
```
Comment (yiguolei) [inline be/src/runtime/runtime_state.h:437]: enable_delete_sub_predicate_v2

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 12]
```diff
@@ -1109,6 +1111,9 @@ public void setMaxJoinNumberOfReorder(int maxJoinNumberOfReorder) {
     @VariableMgr.VarAttr(name = PARALLEL_SYNC_ANALYZE_TASK_NUM)
     public int parallelSyncAnalyzeTaskNum = 2;
 
+    @VariableMgr.VarAttr(name = ENABLE_DELETE_SUB_PREDICATE_V2, fuzzy = true)
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:1114]: forward to master == true??

File: be/src/olap/data_dir.cpp [hunk 13]
```diff
@@ -379,6 +381,17 @@ Status DataDir::load() {
...
+        if (rowset_meta->has_delete_predicate()) {
+            auto* delete_pred = rowset_meta->mutable_delete_pred_pb();
+            DeleteHandler::convert_to_sub_pred_v2(delete_pred, rowset_meta->tablet_schema());
+            LOG(INFO) << fmt::format(
+                    "convert rowset with old delete pred: rowset_id={}, tablet_id={}",
+                    rowset_id.to_string(), tablet_uid.to_string());
+            std::string result;
+            rowset_meta->serialize(&result);
+            std::string key = ROWSET_PREFIX + tablet_uid.to_string() + "_" + rowset_id.to_string();
+            _meta->put(META_COLUMN_FAMILY_INDEX, key, result);
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp:393]: Add a check here: check the field for delete predicate v1 is not changed.

File: be/src/olap/data_dir.cpp [hunk 14]
```diff
@@ -379,6 +381,17 @@ Status DataDir::load() {
         if (rowset_meta->is_local()) {
             rowset_meta->set_fs(local_fs);
         }
+        if (rowset_meta->has_delete_predicate()) {
+            auto* delete_pred = rowset_meta->mutable_delete_pred_pb();
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp:385]: copy rowset meta to a temp filed and then check it at line 392.

File: be/src/olap/delete_handler.cpp [hunk 15]
```diff
@@ -81,35 +89,69 @@ Status DeleteHandler::generate_delete_predicate(const TabletSchema& schema,
...
     if (op == "<") {
         op += "<";
     } else if (op == ">") {
         op += ">";
     }
-    string condition_str;
-    if ("IS" == op) {
-        condition_str = condition.column_name + " " + op + " " + condition.condition_values[0];
-    } else {
+    if ("IS" != op) {
```
Comment (yiguolei) [inline be/src/olap/delete_handler.cpp:147]: add comment here, why do this change??

File: be/src/olap/delete_handler.cpp [hunk 16]
```diff
@@ -81,35 +89,69 @@ Status DeleteHandler::generate_delete_predicate(const TabletSchema& schema,
...
+                      << fmt::format(" {} {} {}", condition.column_name, condition.condition_op,
+                                     condition.condition_values[0]);
         }
     }
     del_pred->set_version(-1);
 
     return Status::OK();
 }
 
-std::string DeleteHandler::construct_sub_predicates(const TCondition& condition) {
```
Comment (yiguolei) [inline be/src/olap/delete_handler.cpp:94]: do not change this method. This method's return value is a condition string. But your new method's return value is ConditionOp. I think you should not change the original logic.

File: be/src/olap/delete_handler.cpp [hunk 17]
```diff
@@ -81,35 +89,69 @@ Status DeleteHandler::generate_delete_predicate(const TabletSchema& schema,
             LOG(INFO) << "store one sub-delete condition. condition name=" << in_pred->column_name()
                       << "condition size=" << in_pred->values().size();
         } else {
-            string condition_str = construct_sub_predicates(condition);
```
Comment (yiguolei) [inline be/src/olap/delete_handler.cpp:84]: not change these lines. 
Just add new lines for delete predicate v2

**Comment by TangSiyang2001** [comment]
run buildall

File: be/src/olap/data_dir.cpp [hunk 18]
```diff
@@ -379,6 +381,21 @@ Status DataDir::load() {
         if (rowset_meta->is_local()) {
             rowset_meta->set_fs(local_fs);
         }
+        if (rowset_meta->has_delete_predicate()) {
+            // copy the delete sub pred v1 to check then
+            auto orig_delete_sub_pred = rowset_meta->delete_predicate().sub_predicates();
+            auto* delete_pred = rowset_meta->mutable_delete_pred_pb();
+            DeleteHandler::convert_to_sub_pred_v2(delete_pred, rowset_meta->tablet_schema());
```
Comment (yiguolei) [inline be/src/olap/data_dir.cpp:388]: If the predicate is already delete pred v2, then you also rewrite the rowset meta. It is useless. And will cause the BE restart time very long.

File: be/src/olap/rowset/rowset_meta_manager.cpp [hunk 19]
```diff
@@ -444,25 +441,35 @@ Status RowsetMetaManager::traverse_rowset_metas(
         std::function<bool(const TabletUid&, const RowsetId&, const std::string&)> const& func) {
     auto traverse_rowset_meta_func = [&func](const std::string& key,
                                              const std::string& value) -> bool {
-        std::vector<std::string> parts;
-        // key format: rst_uuid_rowset_id
```
Comment (yiguolei) [inline be/src/olap/rowset/rowset_meta_manager.cpp:448]: why do you change these code?

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]
LGTM

**Comment by TangSiyang2001** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.44 seconds
 stream load tsv:          536 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162295915 Bytes

File: be/src/olap/delete_handler.cpp [hunk 20]
```diff
@@ -112,6 +147,23 @@ std::string DeleteHandler::construct_sub_predicates(const TCondition& condition)
     return condition_str;
 }
 
+std::string DeleteHandler::trans_op(const std::string& opt) {
+    std::string op = string(opt);
```
[the remark this answers is not in the document]
Comment (yiguolei) [reply be/src/olap/delete_handler.cpp:151]: @gavinchou  Do not use so many static functions. If this function will not be used other place. It is useless to make it as a static function.

**Comment by yiguolei** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.51 seconds
 stream load tsv:          538 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17161895932 Bytes

**Review by dataroaring — APPROVED** [review]
LGTM

File: be/src/olap/delete_handler.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (gavinchou) [reply be/src/olap/delete_handler.cpp:151]: I mean "static free function in delete_handler.cpp", just a free function with `static` qualifier in .cpp file,
and no need to declare it as `DeleteHandler::trans_op()`.
Keep it minimal scope.

**Comment by xiaokang** [comment]
not merge this feature in 2.0

**Pull request merged**